src/storage/v2/property_store.cpp

```diff
@@ -2161,13 +2163,74 @@ std::optional<std::vector<PropertyValue>> PropertyStore::ExtractPropertyValues(
   return WithReader(get_property);
 }
 
+/**
+ * In order to read multiple nested properties with minimal backtracking, we
+ * need a history of where previously seen map properties are stored in the
+ * `PropertyStore` buffer.
+ */
+class ReaderPropPositionHistory {
+ public:
+  // The maximum history depth will always be one less than the maximum
+  // amount of nesting in properties we expect to read. For example, to
+  // read a.b.c.d, we need to store three levels of nesting (`a`, `b`, and `c`).
+  ReaderPropPositionHistory(std::size_t max_size) { history_.reserve(max_size); }
+
+  // Move the reader to a position where the expected leaf property will be
+  // located after the reader.
+  bool ScanToPropertyPathParent(Reader &reader, PropertyPath const &path) {
+    std::span<PropertyId const> parent_map{path.begin(), path.end() - 1};
+
+    auto [prev, next] = r::mismatch(history_, parent_map, {}, &History::property_id);
+
+    if (prev != history_.end()) {
+      reader.SetPosition(prev->property_end);
+      history_.erase(prev, history_.end());
+    }
+
+    for (auto inner_property_id : std::ranges::subrange(next, parent_map.end())) {
+      auto info = FindSpecificPropertyAndBufferInfoMinimal(&reader, inner_property_id);
+      if (info.status != ExpectedPropertyStatus::EQUAL) return false;
+      history_.emplace_back(inner_property_id, info.property_end);
+
+      reader.SetPosition(info.property_begin);
+      auto metadata = reader.ReadMetadata();
+      DMG_ASSERT(metadata, "impossible: metadata has disappeared since `FindSpecificPropertyAndBufferInfoMinimal`");
+      if (!metadata) [[unlikely]]
+        return false;
+      reader.SkipBytes(SizeToByteSize(metadata->id_size) + SizeToByteSize(metadata->payload_size));
+    }
+
+    return true;
+  }
+
+ private:
+  struct History {
+    PropertyId property_id;
+    uint32_t property_end;
+  };
+  std::vector<History> history_;
+};
+
 std::vector<PropertyValue> PropertyStore::ExtractPropertyValuesMissingAsNull(
-    std::span<PropertyId const> ordered_properties) const {
+    std::span<PropertyPath const> ordered_properties) const {
   auto get_properties = [&](Reader &reader) -> std::vector<PropertyValue> {
     auto values = std::vector<PropertyValue>{};
     values.reserve(ordered_properties.size());
 
-    auto const get_value = [](Reader &reader, PropertyId property) { return MatchSpecificProperty(&reader, property); };
+    auto max_history_depth = r::max_element(ordered_properties, {}, r::distance)->size() - 1;
+    ReaderPropPositionHistory history{max_history_depth};
+
+    auto const get_value =
+        [&](Reader &reader,
+            PropertyPath const &path) -> std::pair<ExpectedPropertyStatus, std::optional<PropertyValue>> {
+      if (!history.ScanToPropertyPathParent(reader, path)) {
+        return {ExpectedPropertyStatus::MISSING_DATA, std::nullopt};
```

Comment:
Missing data should be null PropertyValue? Not checked but is this returning correctly?